doc/developer/platform/ux.md

```diff
@@ -772,6 +772,15 @@ running will be hidden from users. Any log messages in `materialized` today that
 are meant for end-users will need to instead make their messages available via
 tables or sources in the system catalog.
 
+### Correctness
+
+Materialize aims to maintain three correctness guarantees.
+
+1. *Materialize is serializable (or strictly serializable) against all internal reads and writes.*
+2. *Materialize respects the explicit or implied event order of its sources.*  In practice this means we reclock sources without transactional semantics (Kafka streams) to reflect the order in which the source emits each event and for sources that do have transactional semantics (Postgres, Debezium) we reclock events to respect the transactional boundaries.
+3. *Materialize provides real time recency guarantees.*  This means if we observe and ingest a transaction finishing at time T and we ask materialize about the current state at time T, we produce the correct answer.  Here T is the correctly reclocked time associated with an event.  Guaranteeing real time recency guarantees means that Materialize may choose to block returning an answer to a question until we have complete certainty about the events at time T.  Blocking may not be desirable in practice, so Materialize makes the behavior optional with explicit syntax so that applications requiring this correctness property can have it while others can ignore it and safely assume non-blocking behavior albeit without any guarantees of correctness.
```

Comment:
@frankmcsherry I see where this definition actually leads readers down an incorrect road.  Is it possible to rely on the reclock timestamp here?  So that we can have T be tied to our internal notion of the event?

Reply:
I think we can use reclocking to translate from one notion of "time" to another, specifically from source "time" to our time. However, I don't think this gives us the recency property. In particular, there are two events: 1. user writes data upstream, and 2. user queries materialize, that I don't think have any specific order, in any time reckoning. To fix that, we have to *introduce an order* by doing something like "block the answer to 2. until we confirm that any strictly preceding 1. have completed".

Maybe differently, no matter how we choose to reclock upstream offsets into MZ times, I can still play a game where I sneak a write in upstream and quickly pose a query, and .. if MZ hasn't heard of the update yet, it is at great risk of violating #3 if it doesn't somehow confirm that there are not updates on the way.

Reply:
> "It should not be possible for a user to commit an update to an upstream source, and not observe that update in a subsequently issued query to Materialize"

I like this, but clearer if you un-negate it, I think!

> If a user observes a successful write to an upstream source, future queries against Materialize _must_ reflect the results of that write (unless the user explicitly requests weaker guarantees).

If you want to tie it to reclocking, I think you have to start from the source timestamp:

> If a user observes a successful write to an upstream source at time *T<sub>s</sub>*, future queries against Materialize that involve that source must occur at a time *T<sub>mz</sub>* such that *T<sub>s</sub>* <= *remap(T<sub>mz</sub>)*.

But, if I understand correctly, I think @frankmcsherry's point is that involving reclocking in this property doesn't buy you much; the property stands on its own as Frank has defined it.

Reply:
Agree on moving away from negate-style language.  Going with this:

This means any write committed by an upstream source is reflected in future queries against Materialize (unless the user explicitly requests weaker guarantees).

Also reworked the language a bit.